docker-machine 0.3.0

rhunter opened the pull request:
A new docker-machine was recently released. This version adds a migration path from using the "boot2docker" package directly (with its CLI tools).

rhunter commented:
Hm, it looks like the automated builds failed because of go-lang package problems. I don't know enough about the Go toolchain nor the Homebrew Jenkins setup to guess which layer is likely to be at fault here. 

The relevant part of one of the build logs:
```
==> Checking out revision 8b9eaf2b6fda23550e09bde1054eeab78e5493bd
==> go install
==> ./bin/godep go build -o docker-machine .
src/github.com/docker/machine/ssh/ssh.go:10:2: cannot find package "github.com/Sirupsen/logrus" in any of:
	/usr/local/Cellar/go/1.4.2/libexec/src/github.com/Sirupsen/logrus (from $GOROOT)
	/private/tmp/docker-machine20150618-43029-b7tyug/machine-0.3.0/Godeps/_workspace/src/github.com/Sirupsen/logrus (from $GOPATH)
	/private/tmp/docker-machine20150618-43029-b7tyug/machine-0.3.0/src/github.com/Sirupsen/logrus
main.go:9:2: cannot find package "github.com/docker/machine/commands" in any of:
	/usr/local/Cellar/go/1.4.2/libexec/src/github.com/docker/machine/commands (from $GOROOT)
	/private/tmp/docker-machine20150618-43029-b7tyug/machine-0.3.0/Godeps/_workspace/src/github.com/docker/machine/commands (from $GOPATH)
	/private/tmp/docker-machine20150618-43029-b7tyug/machine-0.3.0/src/github.com/docker/machine/commands
main.go:10:2: cannot find package "github.com/docker/machine/log" in any of:
	/usr/local/Cellar/go/1.4.2/libexec/src/github.com/docker/machine/log (from $GOROOT)
	/private/tmp/docker-machine20150618-43029-b7tyug/machine-0.3.0/Godeps/_workspace/src/github.com/docker/machine/log (from $GOPATH)
	/private/tmp/docker-machine20150618-43029-b7tyug/machine-0.3.0/src/github.com/docker/machine/log
godep: go exit status 1
```

It built without error on my laptop, where Homebrew's "go" was installed as a dependency (so relatively fresh in regards to a Go environment, I imagine).

dunn commented:
Looks like `github.com/Sirupsen/logrus` needs to be added as a `go_resource`, and the revision for `github.com/docker/machine` needs to be a more recent commit that includes `commands` and `log`.

jpadilla commented:
Couldn't this instead use one of the download from a release? Like:

https://github.com/docker/machine/releases/download/v0.3.0/docker-machine_darwin-amd64

mikemcquaid commented:
> Couldn't this instead use one of the download from a release? Like:

Formulae in Homebrew core are always built from source.

rhunter commented:
I've followed @dunn's  suggestions to the `go_resource` sections, and the Jenkins build seems to be passing now.

Specifically :
  * I've added a `go_resource` for `github.com/Sirupsen/logrus` (using whatever happened to be the tip of `master` at the time.)
  * I changed the revision in `go_resource` for `github.com/docker/machine` to be the revision matching the v0.3.0 tag (matching the formula version).

I don't know much about Go packaging (and especially Go packaging within Homebrew) so I may have missed something. In particular, it seems a bit weird that the install process downloads one copy of the "docker-machine" source code as part of the regular Homebrew fetch process, then download another copy of the "docker-machine" source code (at a particular revision) as a `go_resource`.

dunn commented:
It might be possible to remove the docker-machine `go_resource` and maybe even godep as well, and put all the dependencies (and dependencies of dependencies, and...) into the formula.  The Go build process is very fiddly, though, so it might not work and I'm afraid I don't have time to try at the moment.  This is as far as I got before realizing what I was getting into: https://github.com/Homebrew/homebrew/compare/master...dunn:docker-machine

DomT4 commented:
Merged in https://github.com/Homebrew/homebrew/commit/e0c4b73abdb35a33abe2dcf0e0dc56926cae5593. Thank you for your contribution to Homebrew, we appreciate it! :smiley_cat:.

If we need to make larger changes outside the version bump let's tee that up in a separate PR - The Docker ecosystem is fairly delicate and some formulae being out-of-sync is causing some [problems](https://github.com/Homebrew/homebrew/issues/40908).
